mypy/semanal.py

```diff
@@ -1680,6 +1701,95 @@ def visit_assignment_stmt(self, s: AssignmentStmt) -> None:
                 isinstance(s.rvalue, (ListExpr, TupleExpr))):
             self.add_exports(s.rvalue.items)
 
+    def unwrap_final(self, s: AssignmentStmt) -> None:
+        """Strip Final[...] if present in an assignment.
+
+        This is done to invoke type inference during type checking phase for this
+        assignment. Also, Final[...] desn't affect type in any way, it is rather an
+        access qualifier for given `Var`.
+        """
+        if not s.type or not self.is_final_type(s.type):
+            return
+        assert isinstance(s.type, UnboundType)
+        if len(s.type.args) > 1:
+            self.fail("Final[...] takes at most one type argument", s.type)
+        if not s.type.args:
+            s.type = None
+        else:
+            s.type = s.type.args[0]
+        if len(s.lvalues) != 1 or not isinstance(s.lvalues[0], RefExpr):
+            self.fail("Invalid final declaration, ignoring", s)
+            return
+        lval = s.lvalues[0]
+        assert isinstance(lval, RefExpr)
+        s.is_final_def = True
+        if self.loop_depth > 0:
+            self.fail("Final declarations are prohibited within loops", s)
+        if self.type and self.type.is_protocol:
+            self.msg.protocol_members_cant_be_final(s)
+        if (isinstance(s.rvalue, TempNode) and s.rvalue.no_rhs and
+                not self.is_stub_file and not self.is_class_scope()):
+            self.fail("Final names outside stubs must have a value", s)
+        return
+
+    def check_final_implicit_def(self, s: AssignmentStmt) -> None:
+        """Do basic checks for final declaration on self in __init__.
+
+        Additional re-definition checks are performed by `analyze_lvalue`.
+        """
+        if not s.is_final_def:
+            return
+        lval = s.lvalues[0]
+        assert isinstance(lval, RefExpr)
+        if isinstance(lval, MemberExpr):
+            if not self.is_self_member_ref(lval):
+                self.fail("Final can be only applied to a name or an attribute on self", s)
+                s.is_final_def = False
+                return
+            else:
+                assert self.function_stack
+                if self.function_stack[-1].name() != '__init__':
+                    self.fail("Can only declare final attributes in class body"
+                              " or __init__ method", s)
```

Comment:
Really tiny style nits: Leave out "method" for consistency. Write as "... or __init" or "... or '... or "__init__"'.